src/icd/icdbuilder.cpp

```diff
@@ -0,0 +1,170 @@
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** This Plugin is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with this program.  If not, see <http://www.gnu.org/licenses/>.
+**
+****************************************************************************/
+
+#include "icdbuilder.h"
+
+#include <QDesktopServices>
+#include <QMessageBox>
+#include <QObject>
+#include <QStringList>
+
+#include <coreplugin/icore.h>
+
+#include <projectexplorer/buildconfiguration.h>
+#include <projectexplorer/buildmanager.h>
+#include <projectexplorer/buildsteplist.h>
+#include <projectexplorer/makestep.h>
+#include <projectexplorer/project.h>
+#include <projectexplorer/projectexplorerconstants.h>
+#include <projectexplorer/target.h>
+
+#include <cmakeprojectmanager/cmakebuildstep.h>
+
+#include <asn1acn.h>
+
+using namespace Asn1Acn::Internal::Icd;
+using namespace Asn1Acn::Internal;
+using namespace ProjectExplorer;
+
+namespace {
+
+const QString GENERATED_ICD_PREFIX = "Generated ICD: ";
+const QString CMAKE_TARGETS_KEY = "CMakeProjectManager.MakeStep.BuildTargets";
+const QString ICD_TARGET_NAME = "icdFromAsn1";
+
+void openGeneratedIcd(const QString &line)
+{
+    const auto icdHtml = line.mid(GENERATED_ICD_PREFIX.length()).trimmed();
+    QDesktopServices::openUrl(QUrl::fromLocalFile(icdHtml));
+}
+
+bool icdGenerationFinished(const QString &line)
+{
+    return line.startsWith(GENERATED_ICD_PREFIX);
+}
+
+BuildStep *modifiedMakeBuildStep(BuildStepList *steps)
+{
+    auto step = steps->firstOfType<MakeStep>();
+    if (!step)
+        return nullptr;
+    step->setBuildTarget(ICD_TARGET_NAME, true);
+    return step;
+}
+
+BuildStep *modifiedCMakeBuildStep(BuildStepList *steps)
+{
+    for (auto step : steps->steps()) {
+        auto map = step->toMap();
+        if (map.contains(CMAKE_TARGETS_KEY)) {
+            map[CMAKE_TARGETS_KEY] = ICD_TARGET_NAME;
+            step->fromMap(map);
+            return step;
+        }
+    }
+    return nullptr;
+}
+
+BuildStep *modifiedBuildStep(BuildStepList *steps)
+{
+    auto makeStep = modifiedMakeBuildStep(steps);
+    if (makeStep != nullptr)
+        return makeStep;
+    return modifiedCMakeBuildStep(steps);
+}
+
+std::shared_ptr<BuildConfiguration> prepareClone(BuildConfiguration *buildConfig)
+{
+    auto result = std::shared_ptr<BuildConfiguration>{buildConfig};
+    QObject::connect(BuildManager::instance(),
+                     &BuildManager::buildQueueFinished,
+                     [result](bool) mutable { result.reset(); });
+    return result;
+}
+
+std::shared_ptr<BuildConfiguration> config(Project *project)
```

Comment:
I am not sure about the convention used in this project, but I'm more used to functions containing verbs, like getConfig

Reply:
although it is a Qt-style to name getters using nouns, it's not really a getter, but builder, I've renamed it accordingly